### ISSUE TITLE
Missing background daemon for notifications in android 9 (pie)

### ISSUE THREAD

**Issue by Tentarial** [deltachat/deltachat-android#1008]
- Platform: Custom ROM without GAPPS, Android 9 (Pie)
- Delta Chat Version: 0.510.1
- Expected behavior: Getting notifications
- Actual behavior: no notifications

When someone sends me a message, no notification is displayed. As soon as I enter the app, the message is loaded from the server and the notification is displayed. The logs don't show any errors. I granted all needed permissions to the app and disabled all battery customization for it. As far as I know, you have to run the app constantly active in the background with a daemon or something like that to be able to get notifications without GAPPS with Android Pie and newer. Developers from other Messenger apps already discussed this and developed some workarounds for this issue. Maybe you can have a look at it.

**Comment by severin-lemaignan** [comment]
Same issue here on LineageOS 16 (based on Android 9) with MindTheGapps installed.

**Comment by severin-lemaignan** [comment]
(note that other open-source apps like Conversations, for instance, display the notifications as expected)

**Comment by Pimon1** [comment]
Same issue here. Another example: FOSS-Telegram displays a permanent notification to keep the sync daemon running in the background. It's not nice but it works.
"Works for me Workaround" for now: Subscribe to your Delta Chat IMAP account in K-9 Mail, which shows notifications as expected.

**Comment by r10s** [comment]
can you say if the message is really downloaded later or is just the notification missing?

a look at "settings / advaced / show log" might give some information about the time the message was downloaded. but also a closer look when the app is downloaded: are the "badge counters" in the app already there instantly or are they appearing and increasing when the app is opened?

**Comment by Tentarial** [comment]
@r10s The logs say that the whole IMAP and SMTP sync process and the following notification pop-up only starts when opening the app. You can also see that in the increasing badge counters next to the chats. To make it clear, the app does only behave like that when it is not loaded in the RAM and the task-manager. When you exit the app by just tapping the home button, the app running in the background works perfectly fine.

**Comment by r10s** [comment]
@Tentarial thanks for looking up :)

**Comment by csb0730** [comment]
Got some complaints from two different friends in that way. They see no new message indication at all ! (latest DC version from google play)
At one of these two friends no message appeared until DC app is opened. Then several new messages appeared in two seconds! Badge counter is not displayed (!)

Not shure what Android version they have but will find out ;-)

**Comment by TimoKramer** [comment]
this might be the issue others solved already: https://github.com/Telegram-FOSS-Team/Telegram-FOSS/blob/master/Notifications.md

**Comment by severin-lemaignan** [comment]
Interestingly enough, they mention the change of behaviour since Android 8.0, even though I have no issues with delta.chat notifications on Android 8.0.

Google must have tighten even more the background policies in Android 9.0.

On 07/08/2019 16:00, Timo Kramer wrote:
> this might be the issue others solved already: https://github.com/Telegram-FOSS-Team/Telegram-FOSS/blob/master/Notifications.md
>



-- 
Dr. Séverin Lemaignan
Senior Researcher in Cognitive Robotics and Human-Robot Interaction

Bristol Robotics Laboratory
University of the West of England, Bristol
T Block, Frenchay Campus
Coldharbour Lane
Bristol BS16 1QY

Tel: +44 117 328 5478
Email: severin.lemaignan@brl.ac.uk
Web: https://academia.skadge.org
Skype: severin-lemaignan
Twitter: @skadge

**Comment by r10s** [comment]
yip, we're aware of that. at least in android8, enabling "allow delta to consume battery" does the trick. no idea, why this does not work on android9, i think this needs some investigation.

**Comment by ChrisPHL** [comment]
Hi there,
I've read some documentation about background services and its limitations here: https://developer.android.com/about/versions/oreo/background#services
Since my Android 9 (Pie) behaves almost like this:
> While an app is in the foreground, it can create and run both foreground and background services freely. When an app goes into the background, it has a window of several minutes in which it is still allowed to create and use services. At the end of that window, the app is considered to be idle. At this time, the system stops the app's background services, just as if the app had called the services' Service.stopSelf() methods.

... I wonder if Delta Chat implements it taking these docs into account.

> Google must have tighten even more the background policies in Android 9.0.

Yes, I think they did: 
https://developer.android.com/about/versions/pie/power

It's just a guess from my software developers brain ;-): Delta Chat suffers from the "App Standby Buckets" system. What do you think? At least it could explain, why just some people encounter this issue (low end users) and some not (power users).

kind regards
Christian

**Comment by r10s** [comment]
> ... I wonder if Delta Chat implements it taking these docs into account.

well, there are so many differences between the different android versions, this is hard to follow :)

@ChrisPHL do you know if "disabling battery optimisations" for delta chat allows delta again to get background notifications?

("disabling battery optimisations" is an option somewhere in the system settings, also offered by delta chat directly after the first message is sent) (the current status is shown in delta at settings / advanced / view log)

**Comment by ChrisPHL** [comment]
> @ChrisPHL do you know if "disabling battery optimisations" for delta chat allows delta again to get background notifications?

I assume not, since my log shows "ignoreBatteryOptimizations=true" (Android 9 Pie, sdk=28)

**Comment by r10s** [comment]
okay, thanks, good to know!

/me has to get an android 9 phone ...

**Comment by sp4rkie** [comment]
the funny thing with my XZ1 compact/ Android 9 is: 

when setting ring and notification to non-zero everything works as expected. Even the 
visual notification works then. 

Activating vibration alone is not sufficient i.e. nothing then is notified when deltachat is running in background

**Comment by darkestDark** [comment]
Same Problem as mentioned before on android 9... 
without battery optimisation the new grades of usage doesnt appear so there cant be the problem.
think the problem is because of notification level... created a pr for it...
https://github.com/deltachat/deltachat-android/pull/1042

**Comment by webratte** [comment]
This seems not fixed yet.
On my test phone (Samsung Galaxy S8 Android  9)
Delta Chat: 0.940.1
it dosn't work.

My testmessage was received at 14:15:03 and sent 14:11:55

Logfile:
[Wachtnichtauf.zip](https://github.com/deltachat/deltachat-android/files/3793982/Wachtnichtauf.zip)

**Comment by webratte** [comment]
I just found a interesting option, hidden deep in the system settings of my Samsung

**Adaptiv battery**
Limit battery usage for apps that you don't use often. Your phone will learn how you use apps over time.

This option is enabled by default. 
I have just switched it off and will test a while.

So I will see if it will work again.

**Comment by webratte** [comment]
Ah, thank you.

How ever...
the app seems not to wake up if the phone is in standby for a couple of minutes.

So the initial issue seems not be fixed.
Or maybe only for a few phone models.

**Comment by sp4rkie** [comment]
for  XZ1 compact/ Android 9 the issue seems to be fixed by  version 0.920.0
